File: remove_k_digits/Remove_K_Digits.cpp
class Remove_K_Digits {
public:
    string removeKdigits(string num, int k) {
        if (num.empty() || num.size() <= k) {
            return "0";
        }
        if (k < 1) {
            return num;
        }

        deque<int> stack;
        for (int i = 0; i < num.size(); i++) {
            while (!stack.empty() && stack.front() > num[i] - '0' && k-- > 0) {
                stack.pop_front();
            }
            stack.push_front(num[i] - '0');
        }

        string removed_num = "";
        while (!stack.empty() && (int)stack.size() > k) {
            if (removed_num.empty() && stack.back() == 0) {
                stack.pop_back();
                continue;
            }
            removed_num.push_back((char)(stack.back() + '0'));
            stack.pop_back();
        }
        return removed_num.size() == 0 ? "0" : removed_num;
    }
};

File: remove_k_digits/RemoveKDigits.java
class RemoveKDigits {
    public String removeKdigits(String num, int k) {
        if (num == null || num.length() == 0 || k >= num.length()) {
            return "0";
        }
        if (k < 1) {
            return num;
        }

        Deque<Integer> stack = new ArrayDeque<>();
        for (int i = 0; i < num.length(); i++) {
            int digit = num.charAt(i) - '0';
            while (!stack.isEmpty() && stack.peekFirst() > digit && k-- > 0) {
                stack.pollFirst();
            }
            stack.offerFirst(digit);
        }

        StringBuilder sb = new StringBuilder();
        while (!stack.isEmpty() && stack.size() > k) {
            if (sb.length() == 0 && stack.peekLast() == 0) {
                stack.pollLast();
                continue;
            }
            sb.append(stack.pollLast());
        }
        return sb.isEmpty() ? "0" : sb.toString();
    }
}